infertrade/algos/community/allocations.py

```diff
@@ -94,7 +211,120 @@ def high_low_difference(dataframe: pd.DataFrame, scale: float = 1.0, constant: f
     return dataframe
 
 
-def sma_crossover_strategy(dataframe: pd.DataFrame, fast: int = 0, slow: int = 0) -> pd.DataFrame:
+def level_relationship(dataframe: pd.DataFrame) -> pd.DataFrame:
+    # observations:
+    # does not return a new copy of the df, just alters the original df
+    # does not check for NaNs/infinite in input
+    # does not calculate bid-ask spread
+    # does not fill NaNs
+    # does not change NaNs/infinite to 0 after lag/pct_chg (except for for the first time step/period)
+    # is not able to calculate out_of_sample_erro
+
+    df = dataframe.copy()
+    regression_period = 120
+    minimum_length_to_calculate = regression_period + 1
+    if len(dataframe[PandasEnum.MID.value]) < minimum_length_to_calculate:
+        dataframe[PandasEnum.ALLOCATION.value] = 0.0
+        return df
+
+    calculate_level_relationship(df, regression_period)
+
+    return df
+
+
+def calculate_level_relationship(dataframe: pd.DataFrame, regression_period: int, kelly_fraction: float = 1.0):
+    """Calculates allocations for level relationship."""
+    dataframe[PandasEnum.SIGNAL.value] = dataframe.loc[:, 'research']
+    forecast_period = 100
+    signal_lagged = infertrade.utilities.operations.lag(np.reshape(dataframe[PandasEnum.SIGNAL.value].append(pd.Series([0])).values, (-1, 1)),
+                                                        shift=1) # revert back to manually calculating last row? doing it manually seems awkward, doing it this way seems wasteful, altering the the lag (or other) function seems hacky
+    signal_lagged[0] = [0.0]
+    last_feature_row = signal_lagged[-1:]
+    signal_lagged = signal_lagged[:-1]
+    price_pct_chg = infertrade.utilities.operations.pct_chg(dataframe[PandasEnum.MID.value])
+    price_pct_chg[0] = [0.0]
+
+    calculate_regression_with_kelly_optimum(dataframe, feature_matrix=signal_lagged, last_feature_row=last_feature_row, target_array=price_pct_chg, regression_period=regression_period, forecast_period=forecast_period, kelly_fraction=kelly_fraction)
+
+
+def calculate_regression_with_kelly_optimum(dataframe: pd.DataFrame, feature_matrix: pd.Series, last_feature_row: np.ndarray, target_array: pd.Series, regression_period: int, forecast_period: int, kelly_fraction: float = 1.0):
```

Comment:
We could move this to the operations/utilities file, as it is not a direct allocation rule.